lib/pacemaker/pcmk_rule.c

```diff
@@ -0,0 +1,296 @@
+    if (new_data_set == NULL) {
+        return ENOMEM;
+    }
+
+    pe__set_working_set_flags(new_data_set,
+                              pe_flag_no_counts|pe_flag_no_compat);
+
+    // Populate the working set instance
+
+    // Make our own copy of the given input or fetch the CIB and use that
+    if (input != NULL) {
+        new_data_set->input = copy_xml(input);
+        if (new_data_set->input == NULL) {
+            out->err(out, "Failed to copy input XML");
+            pe_free_working_set(new_data_set);
+            return ENOMEM;
+        }
+
+    } else {
+        int rc = cib__signon_query(NULL, &(new_data_set->input));
+
+        if (rc != pcmk_rc_ok) {
+            out->err(out, "CIB query failed: %s", pcmk_rc_str(rc));
+            pe_free_working_set(new_data_set);
+            return rc;
+        }
+    }
+
+    // Make our own copy of the given crm_time_t object; otherwise
+    // cluster_status() populates with the current time
+    if (date != NULL) {
+        // pcmk_copy_time() guarantees non-NULL
+        new_data_set->now = pcmk_copy_time(date);
+    }
+
+    // Unpack everything
+    cluster_status(new_data_set);
+    *data_set = new_data_set;
+
+    return pcmk_rc_ok;
+}
+
+#define XPATH_NODE_RULE "//" XML_TAG_RULE "[@" XML_ATTR_ID "='%s']"
+
+/*!
+ * \internal
+ * \brief Check whether a given rule is in effect
+ *
+ * \param[in]     data_set  Cluster working set
+ * \param[in]     rule_id   The ID of the rule to check
+ * \param[out]    error     Where to store a rule evaluation error message
+ *
+ * \return Standard Pacemaker return code
+ */
+static int
+eval_rule(pe_working_set_t *data_set, const char *rule_id, const char **error)
+{
+    xmlNodePtr cib_constraints = NULL;
+    xmlNodePtr match = NULL;
+    xmlXPathObjectPtr xpath_obj = NULL;
+    char *xpath = NULL;
+    int rc = pcmk_rc_ok;
+    int num_results = 0;
+
+    *error = NULL;
+
+    /* Rules are under the constraints node in the XML, so first find that. */
+    cib_constraints = pcmk_find_cib_element(data_set->input,
+                                            XML_CIB_TAG_CONSTRAINTS);
+
+    /* Get all rules matching the given ID that are also simple enough for us
+     * to check. For the moment, these rules must only have a single
+     * date_expression child and:
+     * - Do not have a date_spec operation, or
+     * - Have a date_spec operation that contains years= but does not contain
+     *   moon=.
+     *
+     * We do this in steps to provide better error messages. First, check that
+     * there's any rule with the given ID.
+     */
+    xpath = crm_strdup_printf(XPATH_NODE_RULE, rule_id);
+    xpath_obj = xpath_search(cib_constraints, xpath);
+    num_results = numXpathResults(xpath_obj);
+
+    free(xpath);
+    freeXpathObject(xpath_obj);
+
+    if (num_results == 0) {
+        *error = "Rule not found";
+        return ENXIO;
+    }
+
+    if (num_results > 1) {
+        // Should not be possible; schema prevents this
```

Comment:
One gotcha that's easy to forget is that schema validation can be disabled, and then the input can be anything as long as it's legal XML. So while the schema validation is important for the user experience, it can't be relied on by the code.